### PR TITLE
AWS: Implement basic retryer with special handling for throttling errors

### PR DESCRIPTION
**Pull request by jtblin**
Following conversation on #19335, here is an example using the [request.Retryer interface](https://github.com/aws/aws-sdk-go/blob/master/aws/request/retryer.go). I haven't tested yet, just for discussion. 

If we want to go that route, we could also improve the exponential backoff algorithm which is fairly simple right now (taken from the `aws` lib).

Issue #12121 

cc @justinsb

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/21137)

### CODE REVIEW

**Comment by jtblin** [comment]
Related issue: ##20316. Makes me think we should really change the default retry to be less aggressive in **every** case and maybe even less aggressive in case of retry.

**Comment by davidopp** [comment]
ok to test

**Comment by jtblin** [comment]
Thinking more about this, I am more and more convinced that the default backoff algorithm is too aggressive and that using larger delays between retries would help a lot with avoiding hitting AWS API limits.

Here are the delays for each retry with the default implementation:

1. 82ms
1. 164ms
1. 328ms

See https://play.golang.org/p/e98xgnuT9k for the extracted default exponential backoff logic from the go aws-sdk.

TL;DR; there will be 3 attempts in ~500ms. Knowing that the API limits vary per category but I have heard that it's sometimes 5/s, so 3 hits in ~500ms is likely to compound the issue as the [EC2 doc about API errors](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/query-api-troubleshooting.html) indicates that failed retry attempts count toward the API limits, compounding the problem.

> If an API request exceeds the API request rate for its category, the request returns the RequestLimitExceeded error code. **To prevent this error, ensure that your application doesn't retry API requests at a high rate**. You can do this by using care when polling and by using exponential back-off retries.

I am thinking that using a less aggressive algorithm we could change to something like:

```
Request #  RetryInterval (seconds)  Randomized Interval (seconds)
  1          0.75                    [0.375,  1.125]
  2          1.125                   [0.562,  1.687]
  3          1.687                   [0.8435, 2.53]
  4          2.53                    [1.265,  3.795]
  5          3.795                   [1.897,  5.692]
(...)
```

We could have more retries and we could potentially even double the delay in case of throttling errors.

**Comment by justinsb** [comment]
@jtblin I think we want to keep the normal AWS error retry behaviour, in case of transient errors.  So I'm wary of just increasing the delays across-the-board unless we do actually see a rate limit error.

I am concerned also that the first request will also count against our rate limits, and that a delay that works through Retryer doesn't delay that first request.

I'm going to put the alternative #19335 through e2e testing, and see if I can gain confidence in it.  I think we've made huge strides here already in 1.2 (thanks for your help!), and I'd like a final safety belt, but I am wary of changing the AWS behaviour until we've seen the rate limit error, particularly this late in the release process...

**Comment by jtblin** [comment]
Fair enough. I agree there is no exact science here as it is trying to second guess AWS limits opaque behaviour. 

> I am concerned also that the first request will also count against our rate limits, and that a delay that works through Retryer doesn't delay that first request.

That's pretty much the case for everything, until we put a real cluster wide rate limiter in place imo.

> I am wary of changing the AWS behaviour until we've seen the rate limit error, particularly this late in the release process...

Totally makes sense. Do you want me to remove the last commit i.e. only increase the delay if we've hit the delay or shall we just park that now in favour of your other PR?

**Comment by justinsb** [comment]
We merged #19335 so let's see how that goes.  Going to punt this to next-candidate milestone for now

**Comment by jtblin** [comment]
Not sure this PR is needed now that #19335 has been merged although it could be complementary.

@justinsb
